### PR TITLE
Issue 994  empty

### PR DESCRIPTION
**Pull request by kohr-h**

### CODE REVIEW

File: odl/discr/grid.py [hunk 1]
```diff
@@ -237,7 +234,7 @@ def shape(self):
     @property
     def size(self):
         """Total number of grid points."""
-        return np.prod(self.shape)
+        return 0 if self.shape == () else np.prod(self.shape)
```
Comment (adler-j) [inline odl/discr/grid.py]: Not needed:

```
>>> np.prod(())
0.0
```

Comment (kohr-h) [inline odl/discr/grid.py]: I get
```python
>>> np.prod(())
1.0
```

Comment (kohr-h) [inline odl/discr/grid.py]: Not sure if it depends on the Numpy version, but I'll write a comment in the code.

File: odl/space/pspace.py [hunk 2]
```diff
@@ -577,14 +578,28 @@ def __str__(self):
 
     def __repr__(self):
         """Return ``repr(self)``."""
+        # TODO: add the other optional arguments
```
Comment (adler-j) [inline odl/space/pspace.py]: Well, todo :)

File: odl/test/discr/grid_test.py [hunk 3]
```diff
@@ -509,6 +507,48 @@ def test_RectGrid_getitem():
     assert grid[1:3] == sub_grid
 
 
+def test_empty_grid():
+    """Check if empty grids behave as expected and all methods work."""
```
Comment (adler-j) [inline odl/test/discr/grid_test.py]: This whole thing could use a few blank lines for readability

File: odl/test/discr/grid_test.py [hunk 4]
```diff
@@ -509,6 +507,48 @@ def test_RectGrid_getitem():
...
+    assert other.insert(1, grid) == other
+    assert grid.squeeze() == grid
+    assert np.array_equal(grid.points(), np.array([]).reshape((0, 0)))
+    assert grid.corner_grid() == grid
+    assert np.array_equal(grid.corners(), np.array([]).reshape((0, 0)))
+    assert grid.meshgrid == ()
+    assert grid[[]] == grid
+    assert np.array_equal(np.asarray(grid), np.array([]).reshape((0, 0)))
+    assert grid == uniform_grid([], [], ())
+    hash(grid)
```
Comment (adler-j) [inline odl/test/discr/grid_test.py]: At least `hash(grid) == hash(same)` and `hash(grid) != hash(other)` when you're at it.

File: odl/space/pspace.py [hunk 5]
```diff
@@ -577,14 +578,28 @@ def __str__(self):
...
-                                         self.spaces[0], self.size)
+            posargs = [self.spaces[0], self.size]
+            optargs = []
+            oneline = True
         else:
-            inner_str = ', '.join(repr(space) for space in self.spaces)
+            posargs = self.spaces
+            optargs = []
+            argstr = ', '.join(repr(s) for s in self.spaces)
+            oneline = (len(argstr) <= 40 and '\n' not in argstr)
```
Comment (kohr-h) [inline odl/space/pspace.py]: Any comment on this? It's an attempt to keep reasonably short signatures on one line, e.g. `ProductSpace(rn(2), rn(3))`, while putting long stuff on separate lines. The number 40 is somewhat arbitrary, but allows 2 (almost 3) levels of `ProductSpace` not to go beyond 80 chars in total.

Comment (adler-j) [inline odl/space/pspace.py]: I guess it is fine. Ideally we'd like a "global" optimization, but I guess that is not feasible so this is fine.

**Comment by kohr-h** [comment]
Comments fixed, merging after CI

**Pull request merged**